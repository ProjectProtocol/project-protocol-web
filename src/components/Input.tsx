import { kebabCase } from 'lodash-es'
import React from 'react'
import { FloatingLabel, FormControl, FormControlProps } from 'react-bootstrap'

interface IInput extends FormControlProps {
  name?: string
  error?: string
  type?: 'text' | 'email' | 'password' | 'tel' | 'zip'
  controlId?: string
  label: string
}

const Input: React.FC<IInput & FormControlProps> = React.forwardRef<
  HTMLInputElement,
  IInput
>(({ error, controlId, type = 'text', label, ...props }, ref) => {
  return (
    <FloatingLabel label={label} controlId={controlId || kebabCase(label)}>
      <FormControl {...props} type={type} ref={ref} placeholder={label} />
      <FormControl.Feedback type="invalid">{error}</FormControl.Feedback>
    </FloatingLabel>
  )
})

export default Input
