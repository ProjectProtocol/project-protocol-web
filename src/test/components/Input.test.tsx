import { render, screen } from '@testing-library/react'
import { it, expect } from 'vitest'

import Input from 'src/components/Input'

it('matches the snapshot', async () => {
  const r = render(<Input label="My Input" />)
  expect(r.baseElement).toMatchSnapshot()
})

it('renders when given a label', async () => {
  render(<Input label="My Input" />)
  expect(screen.queryByLabelText('My Input')).toBeTruthy()
})

it('renders a text input by default', async () => {
  const r = render(<Input label="My Input" />)
  const el = (await r.baseElement.querySelector('input')) as HTMLInputElement
  expect(el.type).toEqual('text')
})

it('renders error message if present', async () => {
  render(<Input label="My Input" error="Hi I am an error" />)
  expect(screen.getByText('Hi I am an error')).toBeTruthy()
})

it('uses kebabCased label for id by default', async () => {
  const { baseElement } = render(
    <Input label="My Input" error="Hi I am an error" />,
  )
  expect(baseElement.querySelector('#my-input')).toBeTruthy()
})

it('accepts an explicit controlId and overrides default id', async () => {
  const { baseElement } = render(
    <Input
      label="My Input"
      error="Hi I am an error"
      controlId={'hamster-dance1'}
    />,
  )
  expect(baseElement.querySelector('#hamster-dance1')).toBeTruthy()
})
