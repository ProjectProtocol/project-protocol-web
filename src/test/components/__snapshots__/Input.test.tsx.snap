// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="form-floating"
    >
      <input
        class="form-control"
        id="my-input"
        placeholder="My Input"
        type="text"
      />
      <div
        class="invalid-feedback"
      />
      <label
        for="my-input"
      >
        My Input
      </label>
    </div>
  </div>
</body>
`;
